File: src/leetcode/0883.ProjectionAreaOf3DShapes/java/projectionArea.java
class Solution {
    public int projectionArea(int[][] grid) {
        // 三个投影面积之和
        int xy=0;  // x0y平面，只要grid>1即+1
        int xz = 0;  // xoz，
        int yz=0;  //yoz平面，每一列最大值
        int m=grid.length,n=grid[0].length;
        int[] col = new int[n];  // 每一列的最大值
        for(int i=0;i<m;i++) {
            int rowMax = 0;  // 每一行得最大值
            for(int j=0;j<n;j++){
                if(grid[i][j]>0) {
                    xy ++;
                }
                // 记录最大值
                rowMax = Math.max(rowMax, grid[i][j]);
                col[j] = Math.max(col[j], grid[i][j]);
            }
            // 每一行
            xz += rowMax;
        }
        // 每一列得最大值，只有在遍历结束才能确定
        for(int j=0;j<n;j++){
            yz += col[j];
        }
        return xy+xz+yz;
    }
}

File: leetcode/0883.ProjectionAreaOf3DShapes/projectionArea.go
package lt883

/*
883. 三维形体投影面积
在 N * N 的网格中，我们放置了一些与 x，y，z 三轴对齐的 1 * 1 * 1 立方体。

每个值 v = grid[i][j] 表示 v 个正方体叠放在单元格 (i, j) 上。

现在，我们查看这些立方体在 xy、yz 和 zx 平面上的投影。

投影就像影子，将三维形体映射到一个二维平面上。

在这里，从顶部、前面和侧面看立方体时，我们会看到“影子”。

返回所有三个投影的总面积。
*/
/*
正视图：每一行最大值之和；
侧视图：每一列最大值之和；
俯视图：柱子个数；
*/
func projectionArea(grid [][]int) int {
	//定义三个变量 记录三个面的面积
	xy := 0 //  xy面 grid值>1则+1
	xz := 0   //正视图  +grid每一行最大值
	yz := 0 // 侧视图   gird每一列最大值
	col := make([]int, len(grid[0]))  //每一列最大值
	for i:=0;i<len(grid);i++{
		row := 0   //记录每一行最大值 一行一行遍历 所以不需要  row数组
		for j:=0;j<len(grid[i]);j++{
			if grid[i][j] > 0 {
				xy ++
			}
			row = max(row, grid[i][j])
			col[j] = max(col[j], grid[i][j])
		}
		//可以得到每一行的最大值
		xz += row
	}
	//处理侧视图 每一列的最大值
	for i:=0;i<len(col);i++{
		yz += col[i]
	}
	return xy+xz+yz
}
func max(a,b int)int{
	if a>b {
		return a
	}
	return b
}